pkg/ruler/remotequerier.go

```diff
@@ -253,6 +256,28 @@ func (q *RemoteQuerier) query(ctx context.Context, query string, ts time.Time, l
 	return v.Type, v.Result, nil
 }
 
+func (q *RemoteQuerier) sendRequest(ctx context.Context, req *httpgrpc.HTTPRequest) (*httpgrpc.HTTPResponse, error) {
+	// Ongoing request may be cancelled during evaluation due to some transient error or server shutdown,
+	// so we'll keep retrying until we get a successful response or context is eventually cancelled.
+	retryConfig := backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: 2 * time.Second,
+		MaxRetries: maxRequestRetries,
+	}
+	retry := backoff.New(ctx, retryConfig)
+
+	for retry.Ongoing() {
+		resp, err := q.client.Handle(ctx, req)
+		if err != nil {
+			level.Warn(q.logger).Log("msg", "failed to perform remote request", "err", err)
```

Comment:
[nit] A couple of suggestions:
1. I would clarify the error message (see suggestion below), adding also a statement what will retry
2. Log only if will effectively retry it (otherwise the final error is logged by the caller and we don't need it here)

```suggestion
			level.Warn(q.logger).Log("msg", "failed to remotely evaluate query expression, will retry", "err", err)
```